File: packages/core/rules/test/shadow-server/negative.js
// Test cases that should NOT be detected (false positives)

const { Server } = require('@modelcontextprotocol/sdk/server');

// Server with auth
const server = new Server({
  name: "test",
  auth: true
});

// Credentials from environment
const config = {
  apiKey: process.env.API_KEY,
  password: process.env.PASSWORD
};

// Specific CORS origin
res.setHeader("Access-Control-Allow-Origin", "https://example.com");
const corsConfig = { cors: { origin: "https://trusted.com" } };

// TLS verification enabled (default)
const httpsOptions = { rejectUnauthorized: true };

// Debug disabled in production
const appConfig = { debug: false };
if (process.env.NODE_ENV === 'development') {
  config.debug = true;
}

File: packages/core/rules/test/shadow-server/negative.py
# Test cases that should NOT be detected (false positives)

import requests
import os

# Credentials from environment
API_KEY = os.getenv("API_KEY")
PASSWORD = os.environ.get("PASSWORD")

# SSL verification enabled (default)
response = requests.get("https://example.com", verify=True)
response2 = requests.get("https://example.com")  # verify=True is default

# Debug disabled in production
app.debug = False
if os.getenv("ENV") == "development":
    app.run(debug=True)
else:
    app.run(debug=False)
